### DISCUSSION TITLE
記憶體位置

### DISCUSSION THREAD

**Discussion by 107602505** [1101-GS4719-C/Python-Community-GS4719-C#240] [Q&A]
想問之前期中考題中 為何init_tuple_a與init_tuple_b的記憶體位置不同，而a和b的記憶體位置就是相同的?
![image](https://user-images.githubusercontent.com/90819282/147642433-3f05713d-ed1c-4ad8-b9e2-1095c4e14886.png)

**Comment by enoch-lightning** [comment]
你的疑問應該是：init_tuple_a與init_tuple_b，元組(tuple)物件(object)為值不可變的資料型態(型別)，而a與b為整數(int)物件，同樣為值不可變的資料型態。值不可變的物件，若兩個物件的值(values)一樣，那兩個物件的記憶體(RAM)位置應該一樣，那為什麼兩個值一樣的整數物件a與b的記憶體(RAM)位置一樣，但兩個值一樣的元組物件，記憶體(RAM)位置卻不一樣。

**Reply by enoch-lightning** [reply]
我的回答：
值(value)不可變(immutable)的物件(object)，因為值不可變，因此若想讓變數所指向物件的值不一樣，只能創建新物件，也就是：若資料型態為不可變物件，若要改變變數的值，則必須產生具有新的值的物件。而從上面這句話無法推論若資料型態為不可變物件，若兩變數的值一樣，所指向的物件一定是同一個物件。

**Reply by enoch-lightning** [reply]
更詳細的原因我還沒有搞懂，但我目前的想法是這樣，從邏輯去想。

**Reply by enoch-lightning** [reply]
我原本想用下面的程式[3]區塊，來找整數i大到多少時，會讓id(a=i)與id(b=i)不一樣。但，我發現下面程式[3]的id(a=i)與id(b=i)的結果**永遠**會一樣。
![image](https://user-images.githubusercontent.com/86505454/147664749-4fe0ec6d-0c09-40bd-b6c2-341dfa57f97d.png)

**Reply by enoch-lightning** [reply]
原因是i存的是整數物件在RAM記憶體中的位置，而a=i是將i指派給a，就是將i裡面存的整數物件的記憶體位置，複製一份給a。而b=i是將i指派給b，就是將i裡面存的整數物件的記憶體位置，複製一份給b。因此，a與b藉由裡面存的記憶體位置，是一樣的，就一定是找到同一個物件，因此id(a)==id(b)一定為True。

- a=i的=就是「指派」

**Reply by enoch-lightning** [reply]
- data = input()的=也是「指派」

**Reply by enoch-lightning** [reply]
「指派」的英文就是assign。

**Reply by 107602505** [reply]
阿沒錯~謝謝你把我的疑問解釋得更詳細

**Comment by xCk27x** [comment]
老天爺 你這篇差點讓我的python小世界直接崩潰==

底下是我查到的資料:
『
使用python命令列時對於小整數[-5,256]區間內的整數,python會建立**小整數物件池**，這些物件一旦建立，就不會回收，所有新建立的
在這個範圍的整數都是直接參照他即可。所以造成在[-5,256]區間內的整數不同變數只要值相同，參照地址也相同。**此範圍外的整數同樣遵循新建一個變數賦予一個地址**。
』[資料來源](https://it145.com/9/169931.html) <-- 超有用 推薦看一下

意思是說-5~256的整數物件一旦在記憶體中建立後就不會再重新建立了，之後的變數都是把指針放到已建立的物件上面，超過這範圍就是每次定義時都給他新地址再把變數的指針指上去。

字串則是因為特殊的定義方式，所以每個字串都像**小整數物件池**一樣，不會被重新定義。方法類似在每個字串在定義時都會被丟到一個字典裡，要定義一個新字串時會先判斷這個字典裡有沒有這個字串，有的話就把變數指向舊字串，沒有才會建立新字串並把變數指向新字串(但前提是字串中沒有空格，要說為啥請參考[資料來源](https://it145.com/9/169931.html)) 。

**Reply by xCk27x** [reply]
而哪些tuple是建立後就不會重新建立的呢？
我自己試了一下，發現空元組 () 以及 (-5)~(256)這些長度為1的tuple都是不會重新建立的，而長度超過2的tuple好像都沒有這種特性。

**Reply by xCk27x** [reply]
而關於串列list，發現並無list有這種特性(包含空串列 [] )，我想這跟list是可變(mutable)物件有關

**Reply by 107602505** [reply]
謝謝這位同學 看完你的資料來源和解釋就懂了~
